File: 24.10/10월 3주차/num_24060.java
import java.io.*;
import java.util.*;

public class num_24060 {
	static int[] temp;	// 배열 정렬할 임시 배열
	static int count = 0;
	static int K;
	static int result = -1;
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		
		int N = Integer.parseInt(st.nextToken());
		K = Integer.parseInt(st.nextToken());
		
		int[] arr = new int[N];
		temp = new int[N];
		
		st = new StringTokenizer(br.readLine());
		for (int i = 0; i < N; i++) {
			arr[i] = Integer.parseInt(st.nextToken());
		}
		
		mergeSort(arr, 0, arr.length - 1);
		
		System.out.println(result);
	}
	
	static void merge(int[] arr, int left, int mid, int right) {
		int i = left, j = mid + 1, k = 0;
		
		// 배열 정렬
		while (i <= mid && j <= right) {
			if (arr[i] <= arr[j]) {
				temp[k++] = arr[i++];
			}
			else {
				temp[k++] = arr[j++];
			}
		}
		
		// 남은 원소 temp 배열에 추가
		while (i <= mid) {
			temp[k++] = arr[i++];
		}
		
		while (j <= right) {
			temp[k++] = arr[j++];
		}
		
		k = 0;
		i = left;
		
		// temp를 원래 배열인 arr에 복사 및 count에 맞는 원소 찾기
		while (i <= right) {
			count++;
			
			if (count == K) {
				result = temp[k];
				break;
			}
			
			arr[i++] = temp[k++];
		}
	}
	
	private static void mergeSort(int[] arr, int left, int right) {
		if (left < right) {
			int mid = (left + right) / 2;
			
			mergeSort(arr, left, mid);
			mergeSort(arr, mid + 1, right);
			merge(arr, left, mid, right);
		}
	}
}

File: 24.11/11월 1주차/num_24060.py
def mergeSort(arr, p, r) :  # arr 오름차순 정렬
    if p < r :
        q = (p + r) // 2    # q는 p, r의 중간 지점
        mergeSort(arr, p, q)    # 왼쪽 부분 정렬
        mergeSort(arr, q + 1, r)    # 오른쪽 부분 정렬
        merge(arr, p, q, r) # 정렬된 부분정렬 병합

def merge(arr, p, q, r) :
    global count, res
    i = p
    j = q + 1
    temp = []

    while i <= q and j <= r :
        if arr[i] <= arr[j] :
            temp.append(arr[i])
            i += 1
        else :
            temp.append(arr[j])
            j += 1

    # 왼쪽 또는 오른쪽 배열이 남은 경우 
    while i <= q :
        temp.append(arr[i])
        i += 1

    while j <= r :
        temp.append(arr[j])
        j += 1
    
    i = p
    t = 0

    # temp에 저장된 결과를 원래 배열에 저장
    while i <= r :
        arr[i] = temp[t]
        count += 1

        if count == K :
            res = arr[i]
            break

        i += 1
        t += 1

N, K = map(int, input().split())
arr = list(map(int, input().split()))

count = 0
res = -1

mergeSort(arr, 0 , N - 1)
print(res)